cli/parse_args.ts

```diff
@@ -329,52 +329,75 @@ export interface ParseOptions<
   unknown?: (arg: string, key?: string, value?: unknown) => unknown;
 }
 
-interface Flags {
-  bools: Record<string, boolean>;
-  strings: Record<string, boolean>;
-  collect: Record<string, boolean>;
-  negatable: Record<string, boolean>;
-  unknownFn: (arg: string, key?: string, value?: unknown) => unknown;
-  allBools: boolean;
-}
-
 interface NestedMapping {
   [key: string]: NestedMapping | unknown;
 }
 
-const { hasOwn } = Object;
+function isNumberString(x: unknown): boolean {
+  if (typeof x === "number") return true;
+  if (/^0x[0-9a-f]+$/i.test(String(x))) return true;
+  return /^[-+]?(?:\d+(?:\.\d*)?|\.\d+)(e[-+]?\d+)?$/.test(String(x));
+}
+
+function setNested(
+  object: NestedMapping,
+  keys: string[],
+  value: unknown,
+  collect = false,
+) {
+  keys.slice(0, -1).forEach((key) => {
+    object[key] ??= {};
+    object = object[key] as NestedMapping;
+  });
+
+  const key = keys[keys.length - 1];
 
-function get<TValue>(
-  obj: Record<string, TValue>,
-  key: string,
-): TValue | undefined {
-  if (hasOwn(obj, key)) {
-    return obj[key];
+  if (collect) {
+    const v = object[key];
+    if (Array.isArray(v)) {
+      v.push(value);
+      return;
+    }
+
+    value = v ? [v, value] : [value];
   }
+
+  object[key] = value;
 }
 
-function getForce<TValue>(obj: Record<string, TValue>, key: string): TValue {
-  const v = get(obj, key);
-  assertExists(v);
-  return v;
+function hasNested(object: NestedMapping, keys: string[]): boolean {
+  keys = [...keys];
+  const lastKey = keys.pop();
+  if (!lastKey) return false;
+  for (const key of keys) {
+    if (!object[key]) return false;
+    object = object[key] as NestedMapping;
+  }
+  return Object.hasOwn(object, lastKey);
 }
 
-function isNumber(x: unknown): boolean {
-  if (typeof x === "number") return true;
-  if (/^0x[0-9a-f]+$/i.test(String(x))) return true;
-  return /^[-+]?(?:\d+(?:\.\d*)?|\.\d+)(e[-+]?\d+)?$/.test(String(x));
+function aliasIsBoolean(
+  aliasMap: Map<string, Set<string>>,
+  booleanSet: Set<string>,
+  key: string,
+): boolean {
+  const set = aliasMap.get(key);
+  if (set === undefined) return false;
+  for (const alias of set) if (booleanSet.has(alias)) return true;
+  return false;
 }
 
-function hasKey(obj: NestedMapping, keys: string[]): boolean {
-  let o = obj;
-  keys.slice(0, -1).forEach((key) => {
-    o = (get(o, key) ?? {}) as NestedMapping;
-  });
+function isBooleanString(value: string) {
+  return value === "true" || value === "false";
+}
 
-  const key = keys.at(-1);
-  return key !== undefined && hasOwn(o, key);
+function parseBooleanString(value: unknown) {
+  return value !== "false";
 }
 
+const FLAG_REGEXP =
+  /^(?:-(?:(?<doubleDash>-)(?<negated>no-)?)?)(?<key>.+?)(?:=(?<value>.+?))?$/s;
```

Comment:
Argument strings are now parsed with one regexp. Before there were multiple smaller regexps, but the argument string had to be processed multiple times.